shell/platform/android/io/flutter/embedding/engine/android/FlutterFragment.java

```diff
@@ -144,9 +149,62 @@ protected static Bundle createArgsBundle(@Nullable String dartEntrypoint,
     return args;
   }
 
+  @Nullable
+  private FlutterEngine flutterEngine;
+
   public FlutterFragment() {
     // Ensure that we at least have an empty Bundle of arguments so that we don't
     // need to continually check for null arguments before grabbing one.
     setArguments(new Bundle());
   }
+
+  @Override
+  public void onAttach(Context context) {
+    super.onAttach(context);
+
+    // When "retain instance" is true, the FlutterEngine will survive configuration
+    // changes. Therefore, we create a new one only if one does not already exist.
+    if (flutterEngine == null) {
+      createFlutterEngine();
+    }
+  }
+
+  /**
+   * Creates a new FlutterEngine instance.
+   *
+   * Subclasses can instantiate their own {@link FlutterEngine} by overriding
+   * {@link #onCreateFlutterEngine(Context)}.
+   *
+   * Subclasses can alter the {@link FlutterEngine} after creation by overriding
+   * {@link #onFlutterEngineCreated(FlutterEngine)}.
+   */
+  private void createFlutterEngine() {
+    // Create a FlutterEngine to back our FlutterView.
+    flutterEngine = onCreateFlutterEngine(getActivity());
+
+    // Allow subclasses to customize FlutterEngine as desired.
+    onFlutterEngineCreated(flutterEngine);
+  }
+
+  /**
+   * Hook for subclasses to customize the creation of the {@code FlutterEngine}.
+   *
+   * By default, this method returns a standard {@link FlutterEngine} without any modification.
+   */
+  @NonNull
+  protected FlutterEngine onCreateFlutterEngine(@NonNull Context context) {
```

Comment:
The naming of this seems awkward to me.  I think this should just be called `createFlutterEngine`.  Having a non-`void` method called `onCreate` seems unusual.

Reply:
This is really a nit though, feel free to disregard if it fits some other pattern that I'm just not aware of.

Reply:
The convention I was trying to follow here is similar to the `onCreateView()` method in Fragments.  There's a primary method invoked in the superclass, in this case called `createFlutterEngine()`, which then internally invokes a subclass hook, in this case called `onCreateFlutterEngine()`.  Does that make sense?

Reply:
Fair enough